### PR TITLE
Update SPDX Expression Parsing

### PR DESCRIPTION
**Pull request by febuiles**

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -142,7 +144,9 @@ async function readConfigFile(
         ref: pieces.groups.ref
       })
     } else {
-      data = fs.readFileSync(path.resolve(filePath), 'utf-8')
+      // repo-local file
+      const fullPath = path.resolve(__dirname, filePath)
+      data = fs.readFileSync(fullPath).toString()
```
Comment (mrysav) [inline src/config.ts:149]: I notice the `utf-8` was dropped here, was there a particular reason for that?

File: src/config.ts [hunk 2]
```diff
@@ -153,21 +157,26 @@ async function readConfigFile(
...
-      'deny-licenses',
-      'fail-on-scopes',
-      'allow-ghsas',
-      'allow-dependencies-licenses',
-      'deny-packages',
-      'deny-groups'
-    ]
+    data = YAML.parse(configData)
+  } catch (e) {
+    throw new Error(`Can't parse YAML: ${(e as Error).message}`)
```
Comment (mrysav) [inline src/config.ts:164]: Not really sure what it looks like to get an error here or what is "supposed" to happen. But in another language I would wrap the error and it [looks like that might be possible here too, if I'm reading this correctly.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error)

```suggestion
    throw new Error(`Can't parse YAML: ${(e as Error).message}`, { cause: e })
```

Although, that said, is there a reason to catch the error here at all if you're just going to re-throw it?

File: src/config.ts [hunk 3]
```diff
@@ -197,7 +209,7 @@ function parseConfigFile(configData: string): ConfigurationOptionsPartial {
     }
     return data
   } catch (error) {
-    throw error
+    throw new Error(`error validating config: ${error}`)
```
Comment (mrysav) [inline src/config.ts:212]:
```suggestion
    throw new Error(`error validating config: ${(error as Error).message}`)
```

Worth casting to get the message here? Not sure what the output would be if this is an object.

File: src/config.ts (hunk 1 above)
Comment (febuiles) [reply src/config.ts:149]: Fixed in 161d112, ty!

File: src/config.ts (hunk 2 above)
Comment (febuiles) [reply src/config.ts:164]: I like this `cause` thingie, TIL ty! I tested it locally and it's not available in ES6 though 😢 

I introduced this after splitting nested `try` blocks, but looking at the code I don't think we're adding anything to what Node already does so I removed it in c5c5f44.

File: src/config.ts (hunk 3 above)
Comment (febuiles) [reply src/config.ts:212]: The reason to cast is to please the TS type checker, because `error` is only defined inside the `catch` we need to give it some hints so it knows it can call `message` to get the error body. I'm avoiding

File: __tests__/external-config.test.ts [hunk 4]
```diff
@@ -59,38 +55,38 @@ test('it parses options from both sources', async () => {
...
-  setInput('config-file', './__tests__/fixtures/no-licenses-config.yml')
+  setInput('config-file', '../__tests__/fixtures/no-licenses-config.yml')
   const config = await readConfig()
   expect(config.fail_on_severity).toEqual('critical')
 })
 
 test('it raises an error when given an unknown severity in an external config file', async () => {
-  setInput('config-file', './__tests__/fixtures/invalid-severity-config.yml')
+  setInput('config-file', '../__tests__/fixtures/invalid-severity-config.yml')
   await expect(readConfig()).rejects.toThrow()
```
Comment (elireisman) [inline __tests__/external-config.test.ts:83]: 🤔 do these paths require the `../__tests__/` prefix or would a `fixtures/` prefix work too?

File: __tests__/spdx.test.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    expect(spdx.satisfies(license, expr)).toBe(false)
+  })
+
+  test('returns false if no matches are found', async () => {
+    const license = 'ISC'
+    const expr = 'MIT OR GPL-2.0'
+
+    expect(spdx.satisfies(license, expr)).toBe(false)
+  })
+})
```
Comment (elireisman) [inline __tests__/spdx.test.ts:53]: 😍

File: src/config.ts (hunk 3 above)
Comment (elireisman) [reply src/config.ts:207]: ☝️ I think what Mitchell is pointing out is that we're missing the `message` call after the case there?

File: src/spdx.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  try {
+    spdxParse(license)
+    return true
+  } catch (e) {
+    // make a note in the log if the exception
+    // was caused by the expression parser, this
+    // usually means a discrepancy in ESM/CJS.
+    if (e instanceof TypeError) {
+      throw new Error('spdx: invalid expression parser')
+    }
```
Comment (elireisman) [inline src/spdx.ts:20]: 🤔 should the error message instead be `spdx: invalid expression` or something? the parser isn't the problem here, it's the inputs right?

Also under which error conditions is it safe to swallow the error and message here and just return `false` as in line 22?

**Comment by elireisman** [comment]
Couple questions, but this is looking good so far, thanks! 🍻

File: __tests__/external-config.test.ts (hunk 4 above)
Comment (febuiles) [reply __tests__/external-config.test.ts:83]: I'm not entirely sure why (maybe `jest`?), but the path needs to be relative to `src/`, hence the change from `./` to `../`:

```
[FAIL] it accepts an external configuration filename

    Unable to fetch or parse config file: ENOENT: no such file or directory, open '/Users/febuiles/w/dr-action/src/fixtures/no-licenses-config.yml'
```

File: src/spdx.ts (hunk 6 above)
Comment (febuiles) [reply src/spdx.ts:20]: > should the error message instead be spdx: invalid expression or something? the parser isn't the problem here, it's the inputs right?

It's the parser in the sense that `TypeError` is what's raised when there's an error loading the parser. I have not seen these errors outside of local development, where they're common if you forget to `npm i` after bootstrapping the project.

> Also under which error conditions is it safe to swallow the error

None, if this fails we should abort the run.

File: __tests__/external-config.test.ts (hunk 4 above)
Comment (elireisman) [reply __tests__/external-config.test.ts:83]: cc @hmaurer @bteng22 thoughts on this? if its nothing, feel free to ignore 👍

File: src/spdx.ts (hunk 6 above)
Comment (elireisman) [reply src/spdx.ts:20]: ☝️ same here does that sync with your understanding @hmaurer ? we just bail if we can't parse a license, even if there may be vuln pkgs etc. found in this run?

again, I have little context/depth on DR Action at present, so if this is a non-concern, feel free to ignore 👍

**Comment by elireisman** [comment]
👋 I've done a pretty major overhaul of this PR, as the potential behavior changes are broad and the licenses expression checking involved a lot of mocking of the SPDX expression and validity in code where the checks are applied.

Along the way I simplified the use of the new SPDX eval library, updated code that relied on error bubble-ups to short circuit license eval to be more robust in the face of illegal inputs, and added a lot of new unit tests to illustrate how the evaluation and validity checks will behave in practice.

That said, there could be surprises here and I'm no TypeScript expert 😂 so please do review this carefully 🙇 cc @hmaurer @bteng22 @mrysav

File: __tests__/spdx.test.ts [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    }
+  ]
+
+  for (const unit of units) {
+    let got: boolean = spdx.satisfiesAny(unit.candidate, unit.licenses)
+    if (got != unit.expected) {
+      console.log(
+        `failing unit test inputs: candidate(${unit.candidate}) licenses(${unit.licenses})`
+      )
+    }
```
Comment (elireisman) [inline __tests__/spdx.test.ts:67]: ☝️ why log here? allowing expression parsing will make for some counter-intuitive matching results and I wanted the tests to illustrate this going forward, so I tried to set these up in a way that adding more trials would be easy as we address corner cases in the future.

To that end, I tried to install [jest-expect-message](https://www.npmjs.com/package/jest-expect-message) but it did not play nicely with Jest + TS, and after seeking some SME advice from the team, and noting in codesearch that I don't see prior art of this in use in similar Jest + TS projects, I gave up on that.

As-is, when these tests fail, you will see which case failed due to the conditional logging. If someone with more TS than I would like to 🍐 on using that instead, or just convert back to single test cases a la vanilla Jest, let me know 👍

File: package.json [hunk 8]
```diff
@@ -41,6 +41,7 @@
     "zod": "^3.22.3"
   },
   "devDependencies": {
+    "@types/jest": "^29.5.12",
     "@types/node": "^20",
     "@types/spdx-expression-parse": "^3.0.4",
     "@types/spdx-satisfies": "^0.1.1",
```
Comment (elireisman) [inline package.json:47]: moved Jest back into dev dependencies where I think it belongs 👍

File: src/licenses.ts [hunk 9]
```diff
@@ -87,15 +87,19 @@ export async function getInvalidLicenseChanges(
...
-          const found = allow.find(spdxExpression =>
-            spdxSatisfies(license, spdxExpression)
-          )
-          validityCache.set(license, found !== undefined)
+          if (spdx.isValid(license)) {
+            const found = spdx.satisfiesAny(license, allow)
+            validityCache.set(license, found)
+          } else {
+            invalidLicenseChanges.unresolved.push(change)
+          }
```
Comment (elireisman) [inline src/licenses.ts:95]: ☝️ the library is very strict about invalid SPDX inputs, so instead of just bucketing these as "unresolved" based on an error bubble up, we'll now check validity specifically as a pre-step when classifying and resolving licenses 👍

File: src/licenses.ts [hunk 10]
```diff
@@ -161,10 +165,11 @@ const setGHLicenses = async (changes: Change[]): Promise<Change[]> => {
 
   return Promise.all(updatedChanges)
 }
+
 // Currently Dependency Graph licenses are truncated to 255 characters
 // This possibly makes them invalid spdx ids
 const truncatedDGLicense = (license: string): boolean =>
-  license.length === 255 && !isSPDXValid(license)
+  license.length === 255 && !spdx.isValid(license)
```
Comment (elireisman) [inline src/licenses.ts:173]: 🤔 this old codepath could be rethought - just a validity check (required anyway) on our inputs will catch these without the specific truncation checks. I didn't bother in this PR

File: src/spdx.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+export function satisfiesAll(
+  candidateExpr: string,
+  licenses: string[]
+): boolean {
+  try {
+    return spdxlib.satisfiesAll(candidateExpr, licenses)
+  } catch (_) {
+    return false
+  }
+}
```
Comment (elireisman) [inline src/spdx.ts:46]: ☝️ The conditions that throw an error in the new SPDX parsing and eval are very strict, and this made testing and using the library more awkward, so I moved to just return `false` in these cases, and in cases where error bubble up was a sentinel in the logic, I moved to apply an explicit validity check on the inputs instead.

I can go into more detail if desired about why Jest didn't play nice with the error bubble up previously for those that I didn't chat with about this already, but I'm pretty happy with the change so far. Let me know if you disagree or want to sync about it though 👍

File: __tests__/spdx.test.ts (hunk 7 above)
Comment (juxtin) [reply __tests__/spdx.test.ts:67]: Agreed, I'm going to remove that.

File: __tests__/spdx.test.ts [hunk 12]
```diff
@@ -59,17 +59,14 @@ test('satisfiesAny', () => {
...
-      console.log(
-        `failing unit test inputs: candidate(${unit.candidate}) licenses(${unit.licenses})`
-      )
-    }
-    expect(got).toBe(unit.expected)
+    const got: boolean = spdx.satisfiesAny(unit.candidate, unit.licenses)
+    test(`should return ${unit.expected} for ("${unit.candidate}", "${unit.licenses}")`, () => {
+      expect(got).toBe(unit.expected)
+    })
   }
```
Comment (elireisman) [inline __tests__/spdx.test.ts:66]: oh nice change! 😍

File: src/licenses.ts (hunk 9 above)
Comment (elireisman) [reply src/licenses.ts:95]: 🤔 due to this change, we may be able to ditch the `try...catch` blocks entirely

File: src/licenses.ts [hunk 13]
```diff
@@ -87,15 +87,19 @@ export async function getInvalidLicenseChanges(
...
-            spdxSatisfies(license, spdxExpression)
-          )
-          validityCache.set(license, found === undefined)
+          if (spdx.isValid(license)) {
+            const found = spdx.satisfiesAny(license, deny)
+            validityCache.set(license, !found)
+          } else {
+            invalidLicenseChanges.unresolved.push(change)
+          }
         }
```
Comment (elireisman) [inline src/licenses.ts:103]: ⚠️ one more note about the new uses of `satisfiesAny` in these evaluations. This should correctly resolve the problems listed in the support Issues that this PR will close out ...except for [this one](https://github.com/actions/dependency-review-action/issues/263). 

As the new test suites for SPDX expression eval show, we could use `satisfies` in place of `satisfiesAny` to make not only the incoming package license spec parse as an expression (as this PR enables) but also treat our allow/deny list entries as expressions. This would enable DR Action config to set not only single-license entries, but expressions like `MIT AND Apache-2.0` and so on.

But, to resolve the case described in the linked support Issue (incoming package includes an `AND` expression) the end user would still need to supply some duplicative entries in the expression-based allow/deny lists to ensure a good match...

Seems like closing out that last Issue should happen after this PR, and after mulling over some product-facing options for supporting these expressions better 👍

File: __tests__/licenses.test.ts [hunk 14]
```diff
@@ -30,7 +29,7 @@ const rubyChange: Change = {
   name: 'actionsomething',
   version: '3.2.0',
   package_url: 'pkg:gem/actionsomething@3.2.0',
-  license: 'BSD',
+  license: 'BSD-3-Clause',
```
Comment (juxtin) [inline __tests__/licenses.test.ts:32]: Just noting this here—`BSD` is not a valid license type in earlier versions of this action either, but tests used to bypass that check.

File: src/licenses.ts (hunk 13 above)
Comment (juxtin) [reply src/licenses.ts:103]: Thanks @elireisman, the description for this PR indicated that it closes that issue but I was able to confirm that it does not. I removed that line from the description :(

File: src/licenses.ts (hunk 10 above)
Comment (juxtin) [reply src/licenses.ts:173]: Yeah, that's a reasonable take. I'll also leave it for now.

**Comment by juxtin** [comment]
From a code review/testing perspective, I'm pretty happy with this right now. My plan is to do a little bit of manual testing to validate that the linked issues really are resolved, and then I'll get this approved, merged, and released.

**Comment by juxtin** [comment]
> From a code review/testing perspective, I'm pretty happy with this right now. My plan is to do a little bit of manual testing to validate that the linked issues really are resolved, and then I'll get this approved, merged, and released.

I'm glad I didn't rush through this, because this PR doesn't resolve _any_ of the issues that were linked above. That's not to say it's a bad PR or that we shouldn't merge it, but it doesn't address the fundamental causes of the behaviors that people have run into.

I'll break it down issue by issue:
* #670 
    * problem: `Apache-2.0 OR BSL-1.0` should be allowed when only `Apache-2.0` is denied. One of its acceptable licenses is implicitly allowed, so it should be fine.
    * reality: as of this PR, we use `spdx.satisfiesAny('Apache-2.0 OR BSL-1.0', 'Apache-2.0')` to evaluate this. It returns `true`, because 'Apache-2.0 OR BSL-1.0' _does_ satisfy 'Apache-2.0'. This just isn't the semantics that we want.
* #575 
    * problem: `EPL-1.0 OR NOASSERTION OR (EPL-1.0 AND NOASSERTION)` is considered an `Invalid SPDX License`.
    * reality: as of this PR, `NOASSERTION` is _still_ considered an invalid license.
* #635 
    * problem: `LGPL-3.0` license is denied when the deny list only contains `LGPL-2.0, LGPLLR`
    * reality: for reasons I haven't totally figured out yet, this is still the behavior as of this PR. For what it's worth, internally we appear to have a license of `GPL-3.0 AND GPL-3.0-only AND GPL-3.0-or-later AND LGPL-3.0-only` for pygithub, but it is displayed in the action summary as `LGPL-3.0`.

**Review by juxtin — APPROVED** [review]
Despite the fact that this doesn't fix any outstanding bugs, I do still think it moves us in the right direction. The remaining bugs will have to be addressed by other work in the future.
